File: app/src/main/java/org/isoron/uhabits/ui/habits/show/views/ScoreCard.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.ui.habits.show.views;

import android.content.*;
import android.support.annotation.*;
import android.util.*;
import android.widget.*;

import org.isoron.uhabits.*;
import org.isoron.uhabits.models.*;
import org.isoron.uhabits.tasks.*;
import org.isoron.uhabits.utils.*;

import java.util.*;

import butterknife.*;

public class ScoreCard extends RelativeLayout
    implements ModelObservable.Listener
{
    public static final int[] BUCKET_SIZES = { 1, 7, 31, 92, 365 };

    @BindView(R.id.spinner)
    Spinner spinner;

    @BindView(R.id.scoreView)
    ScoreChart chart;

    @BindView(R.id.title)
    TextView title;

    @Nullable
    private Habit habit;

    private int color;

    private int bucketSize;

    public ScoreCard(Context context)
    {
        super(context);
        init();
    }

    public ScoreCard(Context context, AttributeSet attrs)
    {
        super(context, attrs);
        init();
    }

    @OnItemSelected(R.id.spinner)
    public void onItemSelected(int position)
    {
        setBucketSizeFromPosition(position);
    }

    @Override
    public void onModelChange()
    {
        refreshData();
    }

    public void setHabit(@NonNull Habit habit)
    {
        this.habit = habit;
        color = ColorUtils.getColor(getContext(), habit.getColor());
        refreshData();
    }

    @Override
    protected void onAttachedToWindow()
    {
        super.onAttachedToWindow();
        if (habit != null)
        {
            habit.getObservable().addListener(this);
            habit.getScores().getObservable().addListener(this);
        }
    }

    @Override
    protected void onDetachedFromWindow()
    {
        if (habit != null)
        {
            habit.getObservable().removeListener(this);
            habit.getScores().getObservable().removeListener(this);
        }

        super.onDetachedFromWindow();
    }

    @NonNull
    private DateUtils.TruncateField getTruncateField()
    {
        DateUtils.TruncateField field;

        switch (bucketSize)
        {
            case 7:
                field = DateUtils.TruncateField.WEEK_NUMBER;
                break;

            case 365:
                field = DateUtils.TruncateField.YEAR;
                break;

            case 92:
                field = DateUtils.TruncateField.QUARTER;
                break;

            default:
                Log.e("ScoreCard",
                    String.format("Unknown bucket size: %d", bucketSize));
                // continue to case 31

            case 31:
                field = DateUtils.TruncateField.MONTH;
                break;
        }

        return field;
    }

    private void init()
    {
        inflate(getContext(), R.layout.show_habit_strength, this);
        ButterKnife.bind(this);

        int defaultPosition = getDefaultSpinnerPosition();
        setBucketSizeFromPosition(defaultPosition);
        spinner.setSelection(defaultPosition);

        if(isInEditMode())
        {
            spinner.setVisibility(GONE);
            title.setTextColor(ColorUtils.getAndroidTestColor(1));
            chart.setPrimaryColor(ColorUtils.getAndroidTestColor(1));
            chart.populateWithRandomData();
        }
    }

    private int getDefaultSpinnerPosition()
    {
        if(isInEditMode()) return 0;
        return InterfaceUtils.getDefaultScoreSpinnerPosition(getContext());
    }

    private void refreshData()
    {
        if (habit == null) return;

        title.setTextColor(color);
        chart.setPrimaryColor(color);

        new BaseTask()
        {
            @Override
            protected void doInBackground()
            {
                List<Score> scores;

                if (bucketSize == 1) scores = habit.getScores().getAll();
                else scores = habit.getScores().groupBy(getTruncateField());

                chart.setScores(scores);
                chart.setBucketSize(bucketSize);
            }
        }.execute();
    }

    private void setBucketSizeFromPosition(int position)
    {
        if(isInEditMode()) return;

        InterfaceUtils.setDefaultScoreSpinnerPosition(getContext(), position);
        bucketSize = BUCKET_SIZES[position];
        refreshData();
    }
}

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/ui/screens/habits/show/views/ScoreCard.kt
/*
 * Copyright (C) 2016-2025 Álinson Santos Xavier <git@axavier.org>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.core.ui.screens.habits.show.views

import org.isoron.platform.time.LocalDate
import org.isoron.platform.time.TruncateField
import org.isoron.platform.time.getToday
import org.isoron.uhabits.core.models.Habit
import org.isoron.uhabits.core.models.HabitGroup
import org.isoron.uhabits.core.models.PaletteColor
import org.isoron.uhabits.core.models.Score
import org.isoron.uhabits.core.preferences.Preferences
import org.isoron.uhabits.core.ui.views.Theme

data class ScoreCardState(
    val scores: List<Score>,
    val bucketSize: Int,
    val spinnerPosition: Int,
    val color: PaletteColor,
    val theme: Theme
)

class ScoreCardPresenter(
    val preferences: Preferences,
    val screen: Screen
) {
    companion object {
        val BUCKET_SIZES = intArrayOf(1, 7, 31, 92, 365)
        fun getTruncateField(bucketSize: Int): TruncateField {
            return when (bucketSize) {
                1 -> TruncateField.DAY
                7 -> TruncateField.WEEK_NUMBER
                31 -> TruncateField.MONTH
                92 -> TruncateField.QUARTER
                365 -> TruncateField.YEAR
                else -> TruncateField.MONTH
            }
        }

        fun buildState(
            habit: Habit,
            firstWeekday: Int,
            spinnerPosition: Int,
            theme: Theme
        ): ScoreCardState {
            val bucketSize = BUCKET_SIZES[spinnerPosition]
            val today = getToday()
            val oldest = habit.computedEntries.getKnown().lastOrNull()?.date ?: today

            val scores = habit.scores.getByInterval(oldest, today).groupBy { score ->
                truncateDate(getTruncateField(bucketSize), score.date, firstWeekday)
            }.map { (date, scores) ->
                Score(
                    date,
                    scores.map {
                        it.value
                    }.average()
                )
            }.sortedBy {
                it.date
            }.reversed()

            return ScoreCardState(
                color = habit.color,
                scores = scores,
                bucketSize = bucketSize,
                spinnerPosition = spinnerPosition,
                theme = theme
            )
        }

        private fun truncateDate(
            field: TruncateField,
            date: LocalDate,
            firstWeekday: Int
        ): LocalDate {
            // firstWeekday: 1=Sunday, 2=Monday, ..., 7=Saturday
            // DayOfWeek enum: SUNDAY(0), MONDAY(1), ..., SATURDAY(6)
            val firstWeekdayDow = org.isoron.platform.time.DayOfWeek.entries[firstWeekday - 1]
            return when (field) {
                TruncateField.WEEK_NUMBER -> date.startOfWeek(firstWeekdayDow)
                TruncateField.MONTH -> date.startOfMonth()
                TruncateField.QUARTER -> date.startOfQuarter()
                TruncateField.YEAR -> date.startOfYear()
                else -> date
            }
        }

        fun buildState(
            habitGroup: HabitGroup,
            firstWeekday: Int,
            spinnerPosition: Int,
            theme: Theme
        ): ScoreCardState {
            val bucketSize = BUCKET_SIZES[spinnerPosition]
            val today = getToday()
            val oldest = if (habitGroup.habitList.isEmpty) {
                today
            } else {
                habitGroup.habitList.minOf {
                    it.computedEntries.getKnown().lastOrNull()?.date ?: today
                }
            }

            val field = getTruncateField(bucketSize)
            val scores = habitGroup.scores.getByInterval(oldest, today).groupBy {
                DateUtils.truncate(field, it.date, firstWeekday)
            }.map { (date, scores) ->
                Score(
                    date,
                    scores.map {
                        it.value
                    }.average()
                )
            }.sortedBy {
                it.date
            }.reversed()

            return ScoreCardState(
                color = habitGroup.color,
                scores = scores,
                bucketSize = bucketSize,
                spinnerPosition = spinnerPosition,
                theme = theme
            )
        }
    }

    fun onSpinnerPosition(position: Int) {
        preferences.scoreCardSpinnerPosition = position
        screen.updateWidgets()
        screen.refresh()
    }

    interface Screen {
        fun updateWidgets()
        fun refresh()
    }
}
